### ISSUE TITLE
Proposal: Merge PRs (not squash) by default

### ISSUE THREAD

**Issue by rwe** [agda/agda-stdlib#1292]
It's difficult right now to see which PRs were merged or whether the correct version of a PR was the actual thing that got pushed up, and which are included in which (master/experimental/development) branches.

In the presence of frequent squashing, it's easy to get into the habit of dismissing unmerged branches as normal. 

For example, when listing `git branch --no-merged`, branches that were squashed+cherry-picked are listed just like never-merged dev branches. There's no way to tell the difference. *Some* of those may be represented as squashed/cherry-picked commits on `master` or `experimental` or `development`, but it's a big challenge to figure out which are and where; or when seeing those commits, where they came from.

This is probably more apparent if your git config is set up to show PR refs:
```ini
# ~/.config/git/config
[remote "origin"]
	fetch = +refs/pull/*/head:refs/remotes/origin/pr/*
```

Logging the git branches in this repo has become unhelpful, because most of them are unmerged orphans:
```sh
git log --graph --oneline --exclude='*/gh-pages' --all
```

This is exacerbated by the fact that there are multiple long-running branches and occasional pushes directly to those branches.

There is currently a downstream build break (https://github.com/agda/agda/issues/4931) caused by some dropped commits. It's unclear if the wrong branch got committed there, something wasn't properly merged, or something got force-pushed unintentionally. The actual _break_ there is probably not caused by the squashed PRs, but I don't think that it's helped by this workflow.

I'd like to suggest changing the default PR action to "merge" rather than "squash". Does that sound reasonable? Has there been discussion around this before?

**Comment by MatthewDaggitt** [comment]
> I'd like to suggest changing the default PR action to "merge" rather than "squash". Does that sound reasonable? Has there been discussion around this before?

The problem is with not squashing is that a) you can get 20+ commits for a single change to the library which makes it very difficult to track down which commits are related, b) you get a whole lot of noise from commits saying "Fixed missing X" or "Removed whitespace" which is very unhelpful if you're trying to work out which commits are responsible for a particular change in behaviour,  c) you end up making tools like `git bisect` almost unusable as it relies on every commit leaving the repository in a working state which is no longer the case if you use merge.

In summary I feel like one commit per one self-contained functionality change is the right way to go.

The problem you referred to in downstream appears to be a misconfigured `.gitmodule` file in the Agda repo. I agree that we could do better at cleaning up old branches though which sounds like it might make your analysis easier?

**Comment by gallais** [comment]
Note that I am using https://github.com/sindresorhus/refined-github and it automatically
deletes the PR branch whenever I merge a PR on github.

**Comment by rwe** [comment]
Hey, sorry for not responding to this earlier!

As far as `agda-stdlib` goes, you've convinced me. I agree, so will close this issue.

In particular, I agree with this:
> In summary I feel like one commit per one self-contained functionality change is the right way to go.

Browsing through some of the recent PRs in `stdlib` agrees with you that most of the changes here tend to be small enough to be atomic 1 edit → 1 commit → 1 PR, though, so that's all probably irrelevant.


So the following is just food for thought, not pushing a position for the stdlib repo. (Maybe for the core one).

I think there are a couple of assumptions in the scenario you outlined, though I believe you that it's typical.

First, re the atomicity of edits, you might have in mind stream-of-consciousness PRs. Those aren't the norm everywhere, and they're not the kind of PR I was thinking of. I am pro-squash 👍 when messy PRs come along. Commits with "WIP" or "address feedback" or whatever shouldn't be merged directly, if the contributor isn't able/willing to clean them up.

Second: PRs corresponding to a single commit. It depends on that kind of change. Sometimes it makes sense (e.g. agda/agda-stdlib#1232), but often it's not (agda/agda-stdlib#1262). At the more extreme end, some of my PRs to the main Agda repo like (agda/agda#4841, agda/agda#4891) include many small commits, each atomic to improve merges or communication of intent for `git blame` or bisection. But each PR represented a single larger semantic feature-ish encapsulated by the merge. It would have been, in my opinion, harmful to squash those.

There are some practical issues with squashing that might not be obvious.
  * Squashing interacts badly with git's recursive merge strategies. Frequent squashed changes will cause more merge/rebase conflicts than the equivalent merges (of well-structured commits). For example, if two people separately fix a whitespace issue, and one of the people also edits something in that same area, `git` will resolve that merge automatically _iff_ the whitespace and edit operation are separate commits. Probably not a big issue here, but the problem scales with the level of activity. The hassle tends to be pushed down to developers who have to rebase local changes on whatever mess got squashed upstream.
  * In repos with multiple long-lived branches (e.g. `master`/`experimental`), I've seen elsewhere that the unnecessary increase in manual conflict resolution can be a source of confusion and merge mistakes. Especially when the branches can't be strictly cascaded (i.e. all PRs landing in one and then merge down). I don't know if that's a problem here—I guessed it might be when I opened this issue, based on the diffs between the branches, but I wasn't sure.
  * On a community level, as a contributor submitting a pull request, it can suck to see a careful set of changes squashed. That's _especially_ the case if there was any conflict resolution prior to squashing: now the contributor's name is on a commit that might contain changes they might not have endorsed. Doing this also removes any commit signatures, if they're using those. And if they have any branches off of their changes (either locally or stacked PRs), it becomes harder for them to rebase their other work than if the commits were the same.

@gallais I also use refined-github (it's great), but deleted branches aren't necessarily locally pruned or prunable. For example, I frequently have branches e.g. "A, B, C, D, E", some of which may be interdependent or stacked, and a combined a local dev branch. Occasionally I'll submit one of those branches as a PR if I think it's stable. But I frequently have further non-stable changes based on it. Once it's merged and I've pulled, it will show up under `git branch --merged` and I can trivially prune that branch, rebase on top of it, whatever. But if it's squashed, instead it will appear that `master` has a completely unrelated commit doing the same thing. Worse, `rerere` may no longer be able to recognize the diff, and so any resolutions have to be manually redone.

Since I'm just playing around with Agda, it's not a real problem for me personally, but just noting that the merge/squash decision can have effects on that type of workflow.

By the way @MatthewDaggitt I agree with your other points, especially when in cases that PRs are not carefully formed. Given the incredibly slow Agda builds, I've rarely used `git bisect` here; maybe I should. I can see why that might be a problem if the tedium of skipping commits is outweighed by improved bisection granularity.

For simplifying history, I'm used to using `git log --graph --simplify-by-decoration` but I _also_ prefer "merge-only masters": having a branch that is a combination of merges and direct commits definitely complicates things. (Separate aside: GitHub provides _terrible_ default merge messages that only refer to PR urls: I actually suspect that's partly a platform lock-in strategy). Oh well.

**Comment by gallais** [comment]
I guess one thing we could do is systematically push to our own forks of the
stdlib rather than branches of the main repo when we are planning to open a PR.
I had a much easier time switching my workflow to doing that than I expected:

After forking the main repo to your personal account, you can add a new remote
to your local copy of the repo by running

```shell
git remote add MYSELF git@github.com:MYSELF/agda-stdlib
```

Then when you want to start a PR it's the usual

```shell
git checkout -b feature-branch
git add (...)
git commit
```

and finally:

```diff
-git push origin -u feature-branch
+git push MYSELF -u feature-branch
```

You can then open a PR on github by comparing across forks rather than just
picking the branch on the main repo.
